racetrack_commons/racetrack_commons/plugin/engine.py

```diff
@@ -129,6 +130,18 @@ def on_moved(self, event):
         observer.schedule(event_handler, path=Path(self.plugins_dir).as_posix(), recursive=False)
         observer.start()
 
+        # inotify may not work on network filesystems
+        def _check_periodically():
+            while True:
+                time.sleep(5 * 60)
+                try:
+                    compare_changes()
+                except BaseException as e:
+                    log_exception(ContextError('periodic check failure', e))
+
+        threading.Thread(target=_check_periodically, daemon=True).start()
```

Comment:
Is this thread solely for checking that the (network) disk can be read? Wouldn't such problem be visible from the start?

Btw, I already reviewed this, but now forgot how it works - there are many handlers (on_moved etc) that start thread with compare_changes in deamon mode. Won't this cause threads to be created and never finished?

Reply:
> Is this thread solely for checking that the (network) disk can be read? Wouldn't such problem be visible from the start?

NFS disk can be read for sure, but for some reasons it doesn't work with inotify. I'm not sure how to check that, but I suppose it might be quite complex.

> Won't this cause threads to be created and never finished?

That's exactly their purpose - to never finish and to monitor plugins until the end of the container's life.

Reply:
> NFS disk can be read for sure, but for some reasons it doesn't work with inotify. I'm not sure how to check that, but I suppose it might be quite complex. Hence the safety check 

Ah, now I see. So it's additional "manual" check that is run in addition to the ChangesHandler (which is using ionotify). Frankly, if ionotify can't be trusted, and you have to add additional layer just to be sure... I'd just remove ionotify. Cause it's quite some code which is not trivial, and it duplicates with the manual check.

> That's exactly their purpose - to never finish and to monitor plugins until the end of the container's life.

Still don't understand why it won't ultimately cause hundreds of thousands of threads to be spinning in the background, watching the changes :] as every time the LAST_CHANGE_FILE is touched in any way, neverending thread is started.

Reply:
> Still don't understand why it won't ultimately cause hundreds of thousands of threads to be spinning in the background, watching the changes :] as every time the LAST_CHANGE_FILE is touched in any way, neverending thread is started.

Now I see the misunderstanding here. The thread `threading.Thread(target=_check_periodically, daemon=True).start()` is spawned only once, inside `_watch_plugins_changes` function, which is being run at startup, in a `PluginEngine` constructor.

While the other threads like this
```python
        def on_created(self, event):
            if Path(event.src_path).name == LAST_CHANGE_FILE:
                logger.debug(f'file creation notified on {event.src_path}')
                threading.Thread(target=compare_changes, daemon=True).start()
```
are only checking for changes and they finish right after it, so it's not "neverending".

I wanted to keep the observer thread alive in case of error so I moved that to separate threads, but now I see I could just keep it inside try-catch block.

>  if ionotify can't be trusted, and you have to add additional layer just to be sure... I'd just remove ionotify

But if inotify works properly, it's better to have the plugins reloaded immediately right after the change, without the need to wait for the next periodic check cycle.

Reply:
> But if inotify works properly, it's better to have the plugins reloaded immediately right after the change, without the need to wait for the next periodic check cycle.

You could set the period to 10 or 15 sec, which from the perspective of user uploading file on server is almost instant, and it's very light on server. Your decision :)

Reply:
It's tough choice. If it's working with Network Filesystem volume, then it's maybe not so healthy to ping it every 10 seconds due to additional requests over the network.